src/core/mount-setup.c

```diff
@@ -397,9 +397,19 @@ int mount_setup(bool loaded_policy) {
          * nspawn and the container tools work out of the box. If
          * specific setups need other settings they can reset the
          * propagation mode to private if needed. */
-        if (detect_container() <= 0)
-                if (mount(NULL, "/", NULL, MS_REC|MS_SHARED, NULL) < 0)
-                        log_warning_errno(errno, "Failed to set up the root directory for shared mount propagation: %m");
+        if (mount(NULL, "/", NULL, MS_REC|MS_SHARED, NULL) < 0)
+                log_warning_errno(errno, "Failed to set up the root directory for shared mount propagation: %m");
+
+        /* systemd-nspawn sets up an "incoming" directory in
+         * setup_propagate(). It must not be a shared mount, otherwise
+         * mounts cannot be moved (MS_MOVE).
+         *
+         * If "incoming" is not a mountpoint, we were just not started
+         * by nspawn, so it's not an error. */
+        if (path_is_mount_point("/run/systemd/nspawn/incoming", 0) > 0) {
+                if (mount(NULL, "/run/systemd/nspawn/incoming",  NULL, MS_SLAVE, NULL) < 0)
+                        log_warning_errno(errno, "Failed to set up the incoming directory as slave: %m");
+        }
```

Comment:
Yes something like that, hardcoding it is not good especially if it comes from another tool that pid1 does not need... if the logic changes later it will be easy to remove it.